drivers/gl: Use github.com/goxjs/gl instead of github.com/go-gl/gl/v2.1/gl.

shurcooL opened the pull request:
This PR is a continuation of #78. It's different in that it simply modifies the existing driver, rather than adding an second driver.

This PR modifies the gl driver to use `golang.org/x/mobile/gl` package. This package offers a slightly higher level interface, with desktop (OS X, Linux) and mobile (iOS, Android) backends. There is ongoing work to add support for Windows and web browsers.

Resolves #45.
Related to #49.

- The x/mobile/gl provides a universal GL interface which can run on desktop (OS X, Linux), mobile (iOS, Android), with current work to add support for Windows and web browsers.
- Change index buffer data type from []uint32 to []uint16, since it offers enough size, and WebGL implementations do not support []uint32 as readily.
- Remove stFloatMat2x3, stFloatMat2x4, stFloatMat3x2, stFloatMat3x4, stFloatMat4x2 and stFloatMat4x3 since x/mobile/gl does not currently have support for those operations they are unused.
- Remove gl.INVALID_INDEX since x/mobile/gl does not have it.
- Remove support for image.Gray format, since it's unused.
- HACK: Hardcode []uint16 support in newIndexBuffer, since x/mobile/gl has poor support for non-[]uint8 types at this time, only []uint16 is used. Need to improve this.
- HACK: Hardcode []float32 support in newVertexStream since it's the only type used, and x/mobile/gl currently has poor support for other types. Need to improve this.

shurcooL commented:
Hmm, Travis is failing because of:

> Disallowing packages: libgles2-mesa-dev

Any ideas why it's disallowing that package (yet libgl1-mesa-dev was fine)?

shurcooL commented:
Fixed Travis by using the usual sudo apt-get install method of installing packages, since libgles2-mesa-dev is [not whitelisted](https://github.com/travis-ci/apt-package-whitelist/blob/master/ubuntu-precise). Meanwhile, I've opened https://github.com/travis-ci/travis-ci/issues/3518 so once that's resolved, it'll be possible to go back to using apt_packages addon.

crawshaw commented:
Does x/mobile/gl support windows yet? (I can't remember if either Hana or I added it.) I hope it's not to hard to add, but I wouldn't want to break any gxui users on windows.

ben-clayton commented:
I'll give it a try now.

ben-clayton commented:
> Does x/mobile/gl support windows yet?
Apparently so

ben-clayton commented:
[Although I'm not sure how.](https://github.com/golang/mobile/blob/master/gl/gl.go)
I'll continue investigating...

ben-clayton commented:
No, I take it back:

```
# github.com/google/gxui/drivers/gl
drivers\gl\texture.go:94: undefined: gl.Texture```

ben-clayton commented:
> Does x/mobile/gl support windows yet? (I can't remember if either Hana or I added it.) I hope it's not to hard to add, but I wouldn't want to break any gxui users on windows.

So AFAICT, no it doesn't. It doesn't seem trivial to add either. The problem is windows doesn't really have an OpenGL header that's worth using (the system one is OpenGL 1.1, which is pretty worthless). Typically, the drivers support more up to date versions of GL - GXUI has shown that most windows users have a driver that supports at least OpenGL 2.1, but the function pointers need to be obtained using something like [```wglGetProcAddress```](https://msdn.microsoft.com/en-us/library/windows/desktop/dd374386%28v=vs.85%29.aspx).

I'm not sure what the best course of action is for adding Windows support into ```x/mobile/gl```. Really all you're going to end up with is something using GLEW or something that looks very similar to [GLOW](https://github.com/go-gl/gl/blob/master/v2.1/gl/procaddr.go).

sunwangme commented:
why change gl driver？let us focus on ui for desktop first。

ben-clayton commented:
> why change gl driver？let us focus on ui for desktop first。

Don't worry. We won't switch until we're confident that desktop (Windows/OSX/Linux) will continue to work. Getting GXUI working on Android / iOS and WebGL would double the number of supported targets, which would be more than a worthwhile goal.

ben-clayton commented:
@shurcooL, @slimsag, @crawshaw - sorry, I've not had much time to look into this recently.
Given that x/mobile/gl windows support [recently got labelled unplanned](https://github.com/golang/go/issues/9306), perhaps it would be simpler to add Android support to go-gl? It seems rather easy to add [new targets to GLOW](https://github.com/go-gl/gl/blob/master/v2.1/gl/procaddr.go), and I believe it wouldn't take much to do the GetProcAddress for Android.

What do you think?

crawshaw commented:
I have been extremely busy with android/iOS platform bugs, so I won't have time to look at windows support before 1.5. If someone can get GLOW working quickly, makes sense to me.

slimsag commented:
@ben-clayton AFAIK you can generate OpenGL ES 2 bindings with:

```
glow generate -api=gl -version=2.0 -out=./gl/2.0/gl
```

And they already work quite well (I've been using them on Linux desktop, not Android though).

We shouldn't need to add a GetProcAddress for Android as it supports EGL, which means you just need to build the code with `-tags=egl` (we should probably make that the default for arm).

I'll see what I can do later today about making egl the default for arm and generating the ES2 bindings as part of the go-gl/gl repository (will send a PR).

ben-clayton commented:
@slimsag wouldn't you want ```-api=gles2```?
I was playing with GLOW GLES2 generation when [I created this bug](https://github.com/go-gl/glow/issues/60). Aside from the harmless error, the output looked good.

> I'll see what I can do later today about making egl the default for arm and generating the ES2 bindings as part of the go-gl/gl repository (will send a PR).

Awesome, thanks. I confess I haven't tried building and running Go on android yet, but I'm keen to give it a try when I get a chance.

slimsag commented:
@ben-clayton Oops -- yes! I had copied the wrong line from [one of my shell scripts](https://github.com/slimsag/gfx/blob/master/internal/regen.sh). Apologies.

I haven't built any Go code for Android yet either, but I'm excited to see how it plays out.

shurcooL commented:
> Given that x/mobile/gl windows support [recently got labelled unplanned](https://github.com/golang/go/issues/9306#event-277698849), perhaps it would be simpler to add Android support to go-gl?

To answer your question, I need to better understand what the end goal/motivation is. Can you elaborate on that please?

Namely, are we looking to maintain current OS X/Linux/Windows support, but also add Android support? What about Web support?

slimsag commented:
@ben-clayton After talking with @shurcooL a bit over Slack, it appears I've probably misunderstood.

If we go with go-gl/gl here, we'll effectively have to add separate duplicate drivers `gxui/drivers/gl3` (Windows/Linux/OSX) `gxui/drivers/gles2` (Android/iOS) and `gxui/drivers/webgl` (HTML5). It would be difficult compared to what adding Windows support to `/x/mobile/gl` would probably take.

shurcooL commented:
The reason I'm asking for additional clarification here is because at this point, from the information I see above, my thoughts are that no, it would not be simpler to add Android support to go-gl than it would be to add Windows support to `x/mobile/gl`.

The main reason is that `go-gl/gl` is not a universal API with multiple backends (the way x/mobile/gl is). It is a collection of distinct low level specific OpenGL API bindings. So if you import `go-gl/gl/2.1/gl` you get OpenGL 2.1, if you import `go-gl/gl/4.5/gl` you get OpenGL 4.5, and if you import a hypothetical - it does not exist - `go-gl/gles/2.0/gles` package, you get OpenGL ES 2.0 bindings. So then `gxui/drivers/gl` would need to import multiple bindings depending on the environment and do all that. `x/mobile/gl` already has that work done inside, it's just missing Windows support, which is a lot easier to add.

ben-clayton commented:
> The main reason is that go-gl/gl is not a universal API with multiple backends (the way x/mobile/gl is). It is a collection of distinct low level specific OpenGL API bindings. So if you import go-gl/gl/v2.1/gl you get OpenGL 2.1, if you import go-gl/gl/v4.5-core/gl you get OpenGL 4.5 Core Profile, and if you import a hypothetical - it does not exist - go-gl/gles/2.0/gles package, you get OpenGL ES 2.0 bindings.

I understand that. My thought was that we could use the same OpenGL ES 2 bindings on Android and on desktop. I know there are subtle differences between GLES and OpenGL, but for what GXUI uses that difference is not substantial.

Unless there's something I'm missing, this would mean Windows, OSX, Linux and Android would all use this hypothetical GLES 2 go-gl, leaving WebGL as the 'other' target. For that, I was thinking that we could create a simple renderer library to sit between GLES2/WebGL and the gxui/driver. This renderer library would expose the small interface required by GXUI, which in the current form is a handful of GL APIs. We could grow this renderer library as the GXUI driver matures, but I don't see a great deal of additional features for delivering GXUI V1.

shurcooL commented:
> My thought was that we could use the same OpenGL ES 2 bindings on Android and on desktop. I know there are subtle differences between GLES and OpenGL, but for what GXUI uses that difference is not substantial.

Desktop Linux supports OpenGL ES, but OS X (10.10) does not support OpenGL ES. I'm not sure about the status of OpenGL ES support in Windows, but it may be mixed or unsupported. [1] [2]

So by switching to using OpenGL ES bindings, you would lose OS X and possibly Windows support, maintain Linux support, and gain Android support.

Additionally, it would become harder or impossible to add a web backend.

Am I understanding things correctly?

[1] http://www.g-truc.net/post-0457.html
[2] https://github.com/glfw/glfw/blob/5733b0e4cb885c51b4ecb6c41bc2fb2b88b4b407/src/nsgl_context.m#L68-L73

> Unless there's something I'm missing, this would mean Windows, OSX, Linux and Android would all use this hypothetical GLES 2 go-gl

If you have references showing that OS X and Windows have support for OpenGL ES, please point me to it. As far as I know, on desktop side, only Linux has full support for OpenGL ES. Windows and OS X do not support OpenGL ES, only OpenGL.

shurcooL commented:
For reference, `x/mobile/gl` uses OpenGL ES on Linux, but OpenGL on OS X. /cc @crawshaw

```Go
/*
#include <stdlib.h>

#ifdef os_linux
#include <GLES2/gl2.h>
#endif
#ifdef os_darwin_arm
#include <OpenGLES/ES2/glext.h>
#endif
#ifdef os_darwin_amd64
#include <OpenGL/gl3.h>
#endif
*/
import "C"
```

ben-clayton commented:
So, my understanding is this:

* OpenGL ES 2 is an (almost perfect) subset of OpenGL 3.1, but the profile must also support EGL. The key differences are mostly straight-up fewer APIs, but the incompatible differences are [nearly all to do with GLSL](http://stackoverflow.com/a/19801406).
* Given that the exposed functions a simple subset, ```go-gl``` should be able to resolve the GLES 2 function addresses (all with matching signatures) even if the context was built with a OpenGL profile.
* Having a context created with the OpenGL ES 2 profile simply means the driver should conform to the [OpenGL ES 2 spec](https://www.khronos.org/registry/gles/specs/2.0/es_full_spec_2.0.25.pdf). So, the driver should throw errors when calling functions or passing arguments that are not part of the profile, and the driver should handle GLSL slightly differently. Khronos made sure that GLES 2 kept the function signatures the same as desktop GL, even though this resulted in many APIs [requiring arguments to be a fixed constant value](https://www.khronos.org/opengles/sdk/docs/man/xhtml/glCheckFramebufferStatus.xml).

If my understanding is correct, none of what I've described should be a problem for GLES 2 ```go-gl``` function resolving on desktop - the OpenGL ES / OpenGL differences are all problems to be handled externally - i.e. by this renderer abstraction I mentioned.

sunwangme commented:
@ben you are right,just as cocos2d-x do. we need not export all opengles api,we just export gxui api based on window&control&canvas&font. we should do more on layout,advanced control.

shurcooL commented:
I've rebased, and pushed commit cbf18e7 that changes to use a fork of `golang.org/x/mobile/gl` with Windows and browser support added. See https://github.com/goxjs/gl. See last few comments at https://go-review.googlesource.com/#/c/8793/ for rationale why a that CL will not be merged as is (at this time), and therefore a temporary fork makes sense until a new upstream replacement package is available.

Therefore, this PR is now:

> This PR modifies the gl driver to use a fork of `golang.org/x/mobile/gl` package with Windows and browser support added. This package offers a slightly higher level interface, with desktop (OS X, Linux, Windows), mobile (iOS, Android) and desktop, mobile browsers backends.

Because Windows support is preserved with the new package, this PR may be viable to merge, and therefore I think it can be reviewed.

I've tested most samples, everything continues to work for me (and everything should be completely go gettable).

ben-clayton commented:
Hi @shurcooL,
Thank you. I'll try to find time to check this out this weekend. Do you have any instructions on how to try out the WebGL renderer?

shurcooL commented:
Awesome, thank you @ben-clayton!

I will provide instructions on how to do that soon!

shurcooL commented:
I had to do a little work to implement creating fullscreen windows, since you've added a fullscreen sample and I wanted to make it work.

I was able to get it to work in browser (via Fullscreen API), but with a caveat.

http://virtivia.com:27080/3p0e9akjg1yj.mov

The caveat is that it works well when toggling the button with keyboard (tab to select, then enter or space). But not so well with the mouse... because by the time event is processed, the "mouse up" callback has returned and I can only use fullscreen API in user event handlers, so it has to wait until the next user event. Full details at https://github.com/goxjs/glfw/issues/3.

shurcooL commented:
Ok, @ben-clayton, here are precise steps to run gxui samples in the browser via GopherJS.

### Preface

To allow gxui to run in browser, we need support for 3 things: window/context creation, user input, and GL bindings. This PR is to go from OpenGL 2.1 API (via `github.com/go-gl/gl/v2.1/gl` package) to an OpenGL ES 2-like API (via `github.com/goxjs/gl` package). However, this PR does not address window/context creation and user input (currently provided by the `github.com/go-gl/glfw/v3.1/glfw` package).

So, in order for OpenGL programs to run both on desktop and in browser, I've created [`github.com/goxjs/glfw`](https://github.com/goxjs/glfw) package. By using build tags, it simply reuses `github.com/go-gl/glfw/v3.1/glfw` on OS X, Linux and Windows desktop platforms, but it implements window/context creation and user input by using browser APIs (HTML5, Canvas element, Fullscreen API, Mouse Lock API, etc.) in browsers (under [`js` architecture build tag](https://github.com/gopherjs/gopherjs#architecture)). Since this PR doesn't include that change (this PR is strictly about switching from using OpenGL 2.1 to OpenGL ES 2.0 to make it browser/mobile-compatible), you'll need to checkout a different branch. It's exactly one small commit ahead, see shurcooL/gxui@2dab5ce.

### Steps

1. Install the latest GopherJS compiler.

    ```bash
go get -u github.com/gopherjs/gopherjs
```

2. Fetch and checkout my use-goxjs-glfw branch, download dependencies.

    ```bash
cd $GOPATH/src/github.com/google/gxui

# Fetch and checkout my use-goxjs-glfw branch.
git fetch https://github.com/shurcooL/gxui use-goxjs-glfw
git checkout FETCH_HEAD
git branch use-goxjs-glfw

# Go get dependencies (there may be some additional ones under js build tag).
# Note that this is only needed because my work is on a different repository and branch,
# otherwise one could just do go get -u -d -tags js github.com/google/gxui and be done.
go get -u -d -tags js ./...

# May need to switch back to use-goxjs-glfw branch since go get -u may go back to master.
git checkout use-goxjs-glfw
```

3. Use `gopherjs serve` as a convenience way to build Go packages using GopherJS and serve them.

    ```bash
# Run in some terminal and keep it open.
gopherjs serve
```

(This is recommended because it's faster/easier. If you want to actually build the javascript file and bring your own HTML (just one that uses a \<script> tag to run the javascript), then run `gopherjs build github.com/google/gxui/samples/panels` to do that. But then you need to create some html and open it in browser.)

4. Open a WebGL-capable browser and navigate to:

http://localhost:8080/github.com/google/gxui/samples/panels

Change URL to try out different samples.

Thanks and let me know if something doesn't work!

ben-clayton commented:
Thanks for the detailed info @shurcooL.
I gave the Windows build a try, and unfortunately we get a fatal error on start up [initializing gl](https://github.com/goxjs/gl/blob/master/gl_windows.go#L23): 
`gl.Init: glActiveTexture`

I believe the problem is that `gl.Init` must be called after [MakeCurrentContext](https://github.com/go-gl/glfw/blob/master/v3.0/glfw/context.go#L16) as [wglGetProcAddress requires a bound context:](https://msdn.microsoft.com/en-gb/library/windows/desktop/dd374386%28v=vs.85%29.aspx) _When no current rendering context exists or the function fails, the return value is NULL._

However, I can get things running by:

* Adding to [Window in desktop.go](https://github.com/goxjs/glfw/blob/master/desktop.go#L56):

```go
func (w *Window) MakeContextCurrent() {
	w.Window.MakeContextCurrent()
	contextSwitcher.MakeContextCurrent(w)
}

func DetachCurrentContext() {
	glfw.DetachCurrentContext()
	contextSwitcher.MakeContextCurrent(nil)
}
```

* And replacing the dummy [ContextSwitcher in gl_windows.go](https://github.com/goxjs/gl/blob/master/gl_windows.go#L16) with:
```go

// ContextSwitcher is this library's GL context switcher. It facilitates switching
// between GL contexts, satisfying ContextSwitcher interface.
var ContextSwitcher contextSwitcher

type contextSwitcher struct{}

func (contextSwitcher) MakeContextCurrent(context interface{}) {
	if context != nil {
		if _, seen := contexts[context]; !seen {
			// First time this context has been made current. Initialize gl.
			contexts[context] = struct{}{}
			err := gl.Init()
			if err != nil {
				log.Fatalln("gl.Init:", err)
			}
		}
	}
}

var contexts = map[interface{}]struct{}{}
```

This feels a little hacky, but not significantly more than the WebGL context. :) Feel free to ignore this suggestion if you can think of any better solutions.
Cheers.

shurcooL commented:
Thanks for the detailed issue report @ben-clayton!

I had actually tested the Windows backend on OS X, that's why this slipped under my radar. I was a little surprised when I tried placing `gl.Init()` inside of `init()` in the `gl` package and it still worked, but I can see that's not the case for Windows since its wglGetProcAddress requires a bound GL context.

I will absolutely fix this. Thank you for your suggestion, I like it a lot actually. It's interesting how it's quite similar to the `ContextSwitcher` work that needs to happen for WebGL contexts.

I can only think of one other simpler alternative: I could reinstate the `gl.Init()` func and make it a requirement to run it after context has been set, just as is the case with `go-gl/gl/2.1/gl`. It would be a noop on platforms where it's not needed.

However, I think your approach is nicer for two reasons: it makes the user code more simple and eliminates a possible user error. As long as it works in all situations, which I think it would, it's better to do more work inside `glfw` and `gl`. The other reason is that it'd be more consistent with the way the WebGL bindings are initialized via the `ContextSwitcher`, having two different ways to initialize would be suboptimal.

Have you had a chance to try it on OS X, Linux and in the browser yet?

shurcooL commented:
Ok, I've resolved the issue you reported (see goxjs/gl#6 for full details). If you can update dependencies to latest versions, and the new commit at `use-goxjs-glfw` branch of my fork (shurcooL/goxui@395cb0bc0c0febfdbb3f3c029cd917f39d8a4514), I believe it should work. Please let me know your findings.

ben-clayton commented on drivers/gl/vertex_stream.go:
```diff
@@ -5,36 +5,41 @@
 
-	"github.com/go-gl/gl/v2.1/gl"
+	"github.com/goxjs/gl"
+	"golang.org/x/mobile/f32"
```
Can we drop this dependency on "golang.org/x/mobile/f32" ?

shurcooL replied:
Yes, we can if desired. It is only used in this one file for its [`Bytes`](https://godoc.org/golang.org/x/mobile/f32#Bytes) helper func to convert `[]float32` into a corresponding `[]byte` input to `gl.BufferData`.

What are your suggestions on how to do that? Would you prefer I copied the `golang.org/x/mobile/f32` package as a subfolder in `github.com/goxjs/gl`? Or would you like to copy the `Bytes` func into `gxui` itself as an unexported helper?

ben-clayton replied:
Probably the latter for now. Thanks.

ben-clayton commented:
I can confirm Windows and OSX work fine, both in desktop mode and in the browser. I haven't tried linux as I need to dust off my linux-box. I'll try in the coming week.
It's very cool. Good work!

I have a few minor concerns I hope we can address before accepting the PR:
* As I commented above, I'd like to remove the dependency on `golang.org/x/mobile/f32` if possible.
* Unless there's a reason for it, I'd also like to remove the transitive dependency on `honnef.co/go/js/xhr`.
* [Calling `gl.Init()`](https://github.com/goxjs/gl/blob/master/gl_windows.go#L21) every time a context is made current isn't good. You mentioned exactly my concern [on this bug](https://github.com/goxjs/gl/issues/6#issuecomment-107103628): GXUI **can** make frequent switches between contexts when we have multiple windows open. If you have two windows animating, I'm guessing performance is going to suck.

This brings me to the `ContextWatcher`:
* Perhaps we could add a `ContextWatcher.OnCreate(Context)` method to the interface for the [`gl.Init`](https://github.com/goxjs/gl/blob/master/gl_windows.go#L21) logic?
* Super-nit - Why not use the name `ContextWatcher.OnAttach` instead of [`OnBecomeCurrent`](https://github.com/goxjs/glfw/blob/master/glfw.go#L11)?

shurcooL commented:
> It's very cool. Good work!

Thanks! And thanks for reviewing @ben-clayton! I will gladly address and respond to your comments in the near future.

shurcooL commented:
> As I commented above, I'd like to remove the dependency on golang.org/x/mobile/f32 if possible.

It is possible, two ways, which would you prefer? See my reply inline at https://github.com/google/gxui/pull/86#discussion_r31706364.

> Unless there's a reason for it, I'd also like to remove the transitive dependency on honnef.co/go/js/xhr.

There is a good reason for it, but it can be moved elsewhere or replaced with `net/http`. See my detailed response at https://github.com/goxjs/glfw/issues/4#issuecomment-108752326.

> Calling gl.Init() every time a context is made current isn't good.

I've opened https://github.com/goxjs/gl/issues/7 and will work on resolving it there.

> Perhaps we could add a `ContextWatcher.OnCreate(Context)` method to the interface for the `gl.Init` logic?

Can you please elaborate on how you envision it being used? Would it be so that `gl.Init` is called only when a new context is created for the first time, but not called when that context is made current subsequently? If I understand things correctly, this has the potential to cause issues if contexts are very incompatible (e.g., different video card hardware). However, it is unlikely, and it would be the same behavior as most OpenGL Go apps, including `gxui` now - which simply calls `gl.Init` once at startup.

> Super-nit - Why not use the name ContextWatcher.OnAttach instead of OnBecomeCurrent?

Honestly, I had a hard time coming up with good names, and I never thought of `OnAttach`. I can change it easily. However, after thinking more about it, I think I like the current `OnBecomeCurrent` and `OnDetached` names better, because they mirror the language being used:

```Go
func (w *Window) MakeContextCurrent() // MakeContextCurrent makes the context of the window current.
func DetachCurrentContext() // DetachCurrentContext detaches the current context.
```

Also consider http://www.glfw.org/docs/latest/context.html.

> This function **makes** the OpenGL or OpenGL ES context of the specified window **current** on the calling thread. A context can only be **made current** on a single thread at a time and each thread can have only a single current context at a time.
> 
> [in]	window	The window whose context to make current, or NULL to **detach** the current context.

As you can see, the surrounding language is consistently "make context current" and "detach [current] context". The word "attach" is never used, so `OnBecomeCurrent` sounds good to me.

shurcooL commented:
I have addressed all points mentioned above, PTAL @ben-clayton.

> As I commented above, I'd like to remove the dependency on golang.org/x/mobile/f32 if possible.

Done in c4e1bc165b279898d21657273fbd390c08d5a1f0.

> Unless there's a reason for it, I'd also like to remove the transitive dependency on honnef.co/go/js/xhr.

Resolved in https://github.com/goxjs/glfw/issues/4#issuecomment-109212033.

> Calling gl.Init() every time a context is made current isn't good.

Resolved in https://github.com/goxjs/gl/issues/7.

> Super-nit - Why not use the name ContextWatcher.OnAttach instead of OnBecomeCurrent?

See motivation in my previous comment; for consistency I renamed it to `OnMakeCurrent` in https://github.com/goxjs/glfw/pull/5 and https://github.com/goxjs/gl/pull/8.

I also had to add calls to `gl.ContextWatcher.OnMakeCurrent(nil)` in place of where `gl.Init()` was previously, since this PR uses `go-gl/glfw/v3.1/glfw` which does not take on the responsibility of notifying the GL bindings of context changes via `gl.ContextWatcher`. See 834a38e9f7d51ca79038f0e0fb92edf254b34e71. This is because `goxjs/gl` bindings require that (well, on Windows backend only, but that's an internal detail). This will not be needed after switching to `goxjs/glfw` which adds support for browser (remember, this PR is only about changing the GL bindings an necessary changes to go from OpenGL 2.1 API to OpenGL ES 2.1 API).

I think that's all the remaining issues on my side, I'm not aware of any remanning ones. The only thing left from my side on this PR are various shortcuts taken, if you grep for the keyword "HACK" in comments. I'd appreciate suggestions on how to resolve those. It's also fine if you merge as is and clean up those shortcuts yourself in followup commits.

shurcooL commented:
I've documented usage of `goxjs/gl` bindings (related to its `ContextWatcher` mechanism) in https://github.com/goxjs/gl/commit/d4a36dde034f885e1dafae350b058cb763fd1e8a?short_path=04c6e90.

ben-clayton commented on README.md:
```diff
@@ -19,24 +19,17 @@ Dependencies
 In order to build GXUI on linux, you will need the following packages installed:
 
-    sudo apt-get install libxi-dev libxcursor-dev libxrandr-dev libxinerama-dev mesa-common-dev libgl1-mesa-dev libxxf86vm-dev
+    sudo apt-get install libxi-dev libxcursor-dev libxrandr-dev libxinerama-dev mesa-common-dev libgles2-mesa-dev libxxf86vm-dev
```
gles on desktop is not widely supported. See my comments below.

shurcooL replied:
Resolved via goxjs/gl#9 and https://github.com/google/gxui/commit/b022e995a0719c25fd40537792ac4ea0338ef204.

ben-clayton commented:
I finally got around to testing on linux, and unfortunately on my desktop machine I get the following panic on startup:

```
$ $GOPATH/bin/polygon
panic: PlatformError: Failed to retrieve context version string

goroutine 1 [running, locked to thread]:
github.com/go-gl/glfw/v3.1/glfw.acceptError(0xc208067cb8, 0x2, 0x2, 0x0, 0x0)
	/usr/local/google/home/bclayton/src/github.com/go-gl/glfw/v3.1/glfw/error.go:164 +0x145
github.com/go-gl/glfw/v3.1/glfw.CreateWindow(0x320, 0x258, 0x7084d0, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/google/home/bclayton/src/github.com/go-gl/glfw/v3.1/glfw/window.go:264 +0x165
github.com/google/gxui/drivers/gl.newViewport(0xc20801e040, 0x320, 0x258, 0x7084d0, 0x7, 0x0, 0xc208067e80)
	/usr/local/google/home/bclayton/src/github.com/google/gxui/drivers/gl/viewport.go:79 +0x1a4
github.com/google/gxui/drivers/gl.func·019()
	/usr/local/google/home/bclayton/src/github.com/google/gxui/drivers/gl/driver.go:196 +0x66
github.com/google/gxui/drivers/gl.func·012()
	/usr/local/google/home/bclayton/src/github.com/google/gxui/drivers/gl/driver.go:61 +0x2f
github.com/google/gxui/drivers/gl.(*driver).driverLoop(0xc20801e040)
	/usr/local/google/home/bclayton/src/github.com/google/gxui/drivers/gl/driver.go:82 +0x75
github.com/google/gxui/drivers/gl.StartDriver(0x796c40)
	/usr/local/google/home/bclayton/src/github.com/google/gxui/drivers/gl/driver.go:51 +0x2b9
main.main()
	/usr/local/google/home/bclayton/src/github.com/google/gxui/samples/polygon/main.go:84 +0x2a
```

I'm guessing this is the switch from using OpenGL 2.1 to OpenGL ES 2. It is my understanding that GLES on desktop GPUs is not commonly supported. I'm also not sure why we're using OpenGL on Windows and OSX but GLES on linux. Is there a reason for this split?

If you'd like me to try getting goxjs/gl running on linux, I can try and reserve some time this week.

Cheers.

shurcooL commented:
Thanks for testing it out on Linux, I will look into that issue.

> It is my understanding that GLES on desktop GPUs is not commonly supported. I'm also not sure why we're using OpenGL on Windows and OSX but GLES on linux. Is there a reason for this split?

Yes, the reason is very simple. It's how the original `golang.org/x/mobile/gl` package was, and at this time, `goxjs/gl` is simply a fork of x/mobile/ with browser (WebGL, implemented directly rather than through another package) and Windows (via `github.com/go-gl/gl/v2.1/gl` package) back ends added. I simply haven't done anything to the OS X and Linux paths so far.

I don't know exactly why @crawshaw decided to implement Linux desktop support via OpenGL ES, but I have a guess (please confirm or correct me), it's likely because the `x/mobile/gl` package started out, as its name implies, by having OpenGL ES bindings for iOS and Android, and desktop support was added later, likely in a way that was simpler.

> If you'd like me to try getting goxjs/gl running on linux, I can try and reserve some time this week.

Thanks, but it would be very easy for me to fix this by having the current Windows backend (via `go-gl/v2.1/gl` package) also take on Linux and maybe OS X as well. The `go-gl/v2.1/gl` package supports OS X, Linux and Windows well. Do you think that's the way to go?

ben-clayton commented:
> The go-gl/v2.1/gl package supports OS X, Linux and Windows well. Do you think that's the way to go? /cc @slimsag

Yeah, keeping with what we know works sounds like a good approach to me. :)

shurcooL commented:
Sorry, I said "This is needed to resolve google/gxui#86 (comment)." there and that closed this PR, that was unintentional.

> Yeah, keeping with what we know works sounds like a good approach to me. :)

This is done now, see https://github.com/goxjs/gl/pull/9.

I'll be making one followup commit here to update dependencies (no longer need to fetch OpenGL ES libraries).

shurcooL commented:
@ben-clayton, I have resolved the outstanding issues mentioned above, PTAL.

I've tested and gxui samples worked for me on OS X and Linux (Ubuntu 14.04 in a VirtualBox). Although it worked for me on Linux with OpenGL ES backend as well. Please let me know if it fixes your Linux test (where you don't have OpenGL ES support seemingly).

ben-clayton commented:
So linux runs again (yay!), windows does not (boo!).
I'm not too sure what happened during this PR, but somewhere along the way we've lost the changes to import ```"github.com/goxjs/glfw"``` instead of ```"github.com/go-gl/glfw/v3.1/glfw"```.
Fixing up all of these (and changing the `glfw.Init` to take the `gl.ContextWatcher` locally makes windows work again.

dskinner commented:
> It is my understanding that GLES on desktop GPUs is not commonly supported.

On linux, you normally need to be running a mesa driver (intel, nouveau) to get EGL support, otherwise it will likely fail. With the push towards wayland, which currently requires EGL to avoid pulling in X, we should start seeing support from proprietary drivers soon (if it's not there already, haven't been keeping up with it).

shurcooL commented:
> I'm not too sure what happened during this PR, but somewhere along the way we've lost the changes to import `"github.com/goxjs/glfw"` instead of `"github.com/go-gl/glfw/v3.1/glfw"`.

That's intended. I wanted to keep the two PRs separate so that the change to glfw can be reviewed separately. From my [comment](https://github.com/google/gxui/pull/86#issuecomment-109652229) above:

> Reminder, this PR is only about changing the GL bindings an necessary changes to go from OpenGL 2.1 API to OpenGL ES 2.1 API.

However, I would expect Windows to work, because of the change I've done in commit 834a38e9f7d51ca79038f0e0fb92edf254b34e71. From the comment above:

> I also had to add calls to `gl.ContextWatcher.OnMakeCurrent(nil)` in place of where `gl.Init()` was previously, since this PR uses `go-gl/glfw/v3.1/glfw` which does not take on the responsibility of notifying the GL bindings of context changes via `gl.ContextWatcher`. See 834a38e9f7d51ca79038f0e0fb92edf254b34e71. This is because `goxjs/gl` bindings require that (well, on Windows backend only, but that's an internal detail). This will not be needed after switching to `goxjs/glfw` which adds support for browser.

---

> So linux runs again (yay!)

Great!

> windows does not (boo!)

Can you please provide more information, so I can look into it and fix it. I expect it to work with the current PR (and latest versions of dependencies). If there's a panic, can you please post a trace? I'll see if I can setup a Windows machine to test it myself later.

ben-clayton commented:
Sorry, I've had a manic day today and my brain didn't properly context-switch in the 10 minutes I got to test this out. I'm not sure why your `OnMakeCurrent` calls didn't fix the issue, but it's possible I screwed up something when fetching the PR on the windows box. 
I'll have another test tomorrow morning. Like you say, I can't see why this wouldn't work.

Thanks for bearing with me, and sorry for being so slow in replying.

shurcooL commented:
No problem, thank you for spending the time on this PR and helping review it in addition to all the other work you're doing!

shurcooL commented:
Hi David,

I have tested this PR on a Windows machine I was able to get access to, and it worked without any issues. Both the current version `gxui`, as well as this PR branch, worked equally well. This is as I expected.

![image](https://cloud.githubusercontent.com/assets/1924134/8147713/74234b28-122e-11e5-9d85-d2cb6942246d.png)

> I'm not sure why your `OnMakeCurrent` calls didn't fix the issue, but it's possible I screwed up something when fetching the PR on the windows box.
> I'll have another test tomorrow morning. Like you say, I can't see why this wouldn't work.

I recommend using [`gostatus`](https://github.com/shurcooL/gostatus) to confirm that all dependencies are latest versions and not checked out on some non-master branch, etc.

If you run it like this:

```bash
# Show status of all dependencies (recursive) of all gxui packages.
$ go list -f '{{join .Deps "\n"}}' github.com/google/gxui/... | gostatus --stdin -v
```

The output should be exactly this:

```
b    github.com/google/gxui/...
     code.google.com/p/freetype-go/...
     github.com/go-gl/gl/...
     github.com/go-gl/glfw/...
     github.com/goxjs/gl/...
```

The `b` under `github.com/google/gxui/...` is expected because you're likely going to have this PR checked out on a different branch. But all other deps should be with a clean status (up to date, master branch checked out, not a dirty working tree, etc.).

ben-clayton commented:
Tried again, and it is working on windows. Accepting the PR. Thanks :).

shurcooL commented:
Thank you for reviewing and merging!
